File: prototyping/bot.rb
class Bot
  attr_reader :c, :r, :marker
  attr_accessor :landscape

  def initialize(params = {})
    @c = params[:c] || 0
    @r = params[:r] || 0
    @marker = params[:marker] || "X"
    @directions = ["up", "right", "down", "left"]
    @algorithm = params[:algorithm] || "right"
    direction = params[:direction] || @directions.sample
    change_direction! until current_direction == direction
  end

  def next_coords
    dc, dr = 0, 0
    case current_direction
      when "up" then dr = -1
      when "right" then dc = 1
      when "down" then dr = 1
      when "left" then dc = -1
    end
    {c: @c + dc, r: @r + dr}
  end

  def change_direction!
    case @algorithm
    when "right" then @directions.rotate!
    when "random" then @directions.shuffle!
    end
  end

  def current_direction
    @directions.first
  end

  def is_about_to_collide?
    c = next_coords[:c]
    r = next_coords[:r]
    r > @landscape.height - 1 || r < 0 || c > @landscape.width - 1 || c < 0 || @landscape.grid[r][c]
  end

  def move_forward!
    @c = next_coords[:c]
    @r = next_coords[:r]
  end
end

File: js/bot.js
function Bot (params) {
  this.c = params.c;
  this.r = params.r;
  this.directions = ['up', 'right', 'down', 'left'];
  this.initializeDirection(params.direction || _.sample(this.directions));
  this.scale = params.scale || 2;
  this.$element = $('<div class="bot"></div>').css({
    width: this.scale,
    height: this.scale,
    top: this.r,
    left: this.c,
    background: _.sample(["#F6F792", "#77C4D3", "#DAEDE2", "#EA2E49", "#FFFFFF"])
  });
}

Bot.prototype.initializeDirection = function (direction) {
  while (this.currentDirection() !== direction) {
    this.changeDirection();
  }
};

Bot.prototype.currentDirection = function () {
  return this.directions[0];
};

Bot.prototype.changeDirection = function () {
  this.directions.rotate();
};

Bot.prototype.nextCoords = function () {
  var dc = 0;
  var dr = 0;
  switch (this.currentDirection()) {
    case 'up':    dr = -1; break;
    case 'right': dc =  1; break;
    case 'down':  dr =  1; break;
    case 'left':  dc = -1; break;
  }
  return {c: this.c + dc, r: this.r + dr};
};

Bot.prototype.isAboutToCollide = function () {
  var coords = this.nextCoords();
  return !_.inRange(coords.r, this.landscape.height) ||
         !_.inRange(coords.c, this.landscape.width)  ||
         this.landscape.grid[coords.r][coords.c]
};

Bot.prototype.moveForward = function () {
  var coords = this.nextCoords();
  this.c = coords.c;
  this.r = coords.r;
  this.$element.css({
    top: this.r * this.scale,
    left: this.c * this.scale
  });
};

Bot.prototype.render = function () {
  this.landscape.$element.append(this.$element);
};
